### PR TITLE
Adding advanced typography capability #297

### PR DESCRIPTION
**Pull request by codecracker2014**
So far OpenPDF doesn't support advanced typography features like processing of GSUB table. Good documentation for typography is available at following link https://docs.microsoft.com/en-us/typography/opentype/spec/, https://docs.microsoft.com/en-us/typography/opentype/spec/gsub

Apache FOP supports advanced typography(https://xmlgraphics.apache.org/fop/1.1/releaseNotes_1.1.html) and is pretty stable. I've added support of advanced typography by utilizing Apache FOP. With this change OpenPDF will be supporting TrueTypeFonts and advanced typography.

### CODE REVIEW

**Comment by andreasrosdal** [comment]
This is an awesome contribution. Is it possible to make the dependency on FOP optional?

**Comment by asturio** [comment]
@codecracker2014 Do you have a sample, using this new advances typography? This would be very nice.

File: openpdf/pom.xml [hunk 1]
```diff
@@ -54,7 +54,16 @@
             <version>${hamcrest.version}</version>
             <scope>test</scope>
         </dependency>
-
+        <dependency>
```
Comment (mluppi) [inline openpdf/pom.xml]: Could you move these new dependencies above the ones with scope `test`?

**Comment by codecracker2014** [comment]
@andreasrosdal thanks for review, i'll update soon.

**Comment by codecracker2014** [comment]
> 
> 
> @codecracker2014 Do you have a sample, using this new advances typography? This would be very nice.

@asturio junit test would be fine?

**Comment by asturio** [comment]
> > @codecracker2014 Do you have a sample, using this new advances typography? This would be very nice.
> 
> @asturio junit test would be fine?

Yeah, junit is fine.

File: openpdf/pom.xml (hunk 1 above)
Comment (asturio) [inline openpdf/pom.xml]: Please set these dependecies as optional, and move them above of the <scope>test</scope>-Dependecies.

Comment (codecracker2014) [reply openpdf/pom.xml]: I've updated dependencies.
I've verified output pdf with hindi language, following is the output:
Pre
image

Post
image

I'm facing couple of challenges with junit test, need your help.
1. Licensing of Font files for hindi or any other language. For ex https://docs.microsoft.com/en-us/typography/font-list/nirmala-ui
2. I wanted to verify PDF output in test case but seems PdfTextExtractor doesn't support characters other than english, please refer code snippet below. 

    `

            @Test
            public void testGlyphSubstitution() throws Exception{
                FontFactory.register("C:\\Windows\\Fonts\\NIRMALA.TTF");
                ByteArrayOutputStream fileBytes = new ByteArrayOutputStream();
                try (Document document = new Document();) {
                PdfWriter.getInstance(document, fileBytes);
                document.open();
                document.add(new Chunk("नमस्ते", FontFactory.getFont("nirmala ui", "Identity-H", false, 10, 0, null)));
                document.close();
                PdfReader pdfReader = new PdfReader(fileBytes.toByteArray());
                PdfTextExtractor pdfTextExtractor = new PdfTextExtractor(pdfReader);
                //Output is empty 
                System.out.println(pdfTextExtractor.getTextFromPage(0));
                System.out.println(pdfTextExtractor.getTextFromPage(1));

            }

`

**Comment by andreasrosdal** [comment]
Will OpenPDF still work correctly, compile and run without crashing, when these optional dependencies are not present?

**Comment by codecracker2014** [comment]
@andreasrosdal i thought optional dependency is to be marked just to avoid version conflicts. If it is required to make this feature optional then this needs to be handled in code.

**Comment by andreasrosdal** [comment]
Yes, I think this feature should be optional, if possible. So that OpenPDF can be used without requiring these extra two dependencies.

**Comment by codecracker2014** [comment]
@andreasrosdal I've made this feature optional based on presence of Fop on classpath.

**Comment by codecracker2014** [comment]
@asturio I've added unit test for typography please review.
For hindi i've used jaldi font(https://github.com/Omnibus-Type/Jaldi) which is having OFL 1.1 license(SIL Open Font License 1.1).

**Comment by codecracker2014** [comment]
I've updated wiki for this feature have a look https://github.com/LibrePDF/OpenPDF/wiki/Supporting-multi-byte-character-languages-like-Hindi-with-TTF-fonts

**Pull request merged**